mlir/test/Dialect/ArmSME/tile-allocation-liveness.mlir

```diff
@@ -430,3 +430,133 @@ func.func @cond_branch_with_backedge(%slice: vector<[4]xf32>) {
+  "test.some_use"(%tileB) : (vector<[4]x[4]xf32>) -> ()
+  cf.br ^bb3
+^bb2:
+  "test.dummy"(): () -> ()
+  "test.dummy"(): () -> ()
+  "test.dummy"(): () -> ()
+  "test.some_use"(%tileA) : (vector<[4]x[4]xf32>) -> ()
+  cf.br ^bb3
+^bb3:
+  return
+}
+
+// -----
+
+//  CHECK-LIVE-RANGE-LABEL: @holes_in_tile_liveness_inactive_overlaps
+//        CHECK-LIVE-RANGE: ========== Coalesced Live Ranges:
+//        CHECK-LIVE-RANGE: ^bb0:
+//        CHECK-LIVE-RANGE: S  arm_sme.get_tile
+//        CHECK-LIVE-RANGE: E  cf.cond_br
+//        CHECK-LIVE-RANGE: ^bb1:
+//        CHECK-LIVE-RANGE:  S arm_sme.get_tile
+//        CHECK-LIVE-RANGE:  | test.dummy
+//        CHECK-LIVE-RANGE:  | test.some_use
+//        CHECK-LIVE-RANGE:  | arm_sme.copy_tile
+//        CHECK-LIVE-RANGE:  E cf.br
+//        CHECK-LIVE-RANGE: ^bb2:
+//        CHECK-LIVE-RANGE: |  test.dummy
+//        CHECK-LIVE-RANGE: |  test.dummy
+//        CHECK-LIVE-RANGE: |  test.dummy
+//        CHECK-LIVE-RANGE: |S arm_sme.get_tile
+//        CHECK-LIVE-RANGE: E| test.some_use
+//        CHECK-LIVE-RANGE:  | arm_sme.copy_tile
+//        CHECK-LIVE-RANGE:  E cf.br
+//        CHECK-LIVE-RANGE: ^bb3:
+//        CHECK-LIVE-RANGE:  E test.some_use
+//        CHECK-LIVE-RANGE:    func.return
+
+// This tests an edge case in inactive live ranges. The first live range is
+// inactive at the start of ^bb1. If the tile allocator did not check if the
+// second live range overlapped the first it would wrongly re-use tile ID 0
+// (as the first live range is inactive so tile ID 0 is free). This would mean
+// in ^bb2 two overlapping live ranges would have the same tile ID (bad!).
+
+// CHECK-LABEL: @holes_in_tile_liveness_inactive_overlaps
+func.func @holes_in_tile_liveness_inactive_overlaps(%cond: i1) {
+  // CHECK: arm_sme.get_tile {tile_id = 0 : i32}
+  %tileA = arm_sme.get_tile : vector<[4]x[4]xf32>
+  cf.cond_br %cond, ^bb2, ^bb1
+^bb1:
+  // CHECK: arm_sme.get_tile {tile_id = 1 : i32}
+  %tileB = arm_sme.get_tile : vector<[4]x[4]xf32>
+  "test.dummy"(): () -> ()
+  "test.some_use"(%tileB) : (vector<[4]x[4]xf32>) -> ()
+  cf.br ^bb3(%tileB: vector<[4]x[4]xf32>)
+^bb2:
+  "test.dummy"(): () -> ()
+  "test.dummy"(): () -> ()
+  "test.dummy"(): () -> ()
+  // CHECK: arm_sme.get_tile {tile_id = 1 : i32}
+  %tileC = arm_sme.get_tile : vector<[4]x[4]xf32>
+  "test.some_use"(%tileA) : (vector<[4]x[4]xf32>) -> ()
+  cf.br ^bb3(%tileC: vector<[4]x[4]xf32>)
+^bb3(%tile: vector<[4]x[4]xf32>):
+  "test.some_use"(%tile) : (vector<[4]x[4]xf32>) -> ()
+  return
+}
+
+// -----
+
+// This is the same as the previous example, but changes the tile types to
+// vector<[16]x[16]xi8>. This means in bb1 the allocator will need to spill the
+// first live range (which is inactive).
+
+// Note: The live ranges are the same as the previous example (so are not checked).
+
+// CHECK-LABEL: @spill_inactive_live_range
+func.func @spill_inactive_live_range(%cond: i1) {
+  // CHECK: arm_sme.get_tile {tile_id = 16 : i32}
```

Comment:
Why not `tile_id = 0`?

Reply:
Because it was spilled (which is the purpose of this test)

Reply:
I don't fully follow this example :) So I modified it like this (updated `^bb0):
```mlir
  func.func @spill_inactive_live_range(%cond: i1) {
    // CHECK: arm_sme.get_tile {tile_id = 16 : i32}
    %tileA = arm_sme.get_tile : vector<[16]x[16]xi8>
   "test.some_use"(%tileA) : (vector<[16]x[16]xi8>) -> ()
    cf.cond_br %cond, ^bb2, ^bb1
  ^bb1:
    // CHECK: arm_sme.get_tile {tile_id = 0 : i32}
    %tileB = arm_sme.get_tile : vector<[16]x[16]xi8>
    "test.dummy"(): () -> ()
    "test.some_use"(%tileB) : (vector<[16]x[16]xi8>) -> ()
    cf.br ^bb3(%tileB: vector<[16]x[16]xi8>)
  ^bb2:
    "test.dummy"(): () -> ()
    "test.dummy"(): () -> ()
    "test.dummy"(): () -> ()
    // CHECK: arm_sme.get_tile {tile_id = 0 : i32}
    %tileC = arm_sme.get_tile : vector<[16]x[16]xi8>
    "test.some_use"(%tileA) : (vector<[16]x[16]xi8>) -> ()
    cf.br ^bb3(%tileC: vector<[16]x[16]xi8>)
  ^bb3(%tile: vector<[16]x[16]xi8>):
    "test.some_use"(%tile) : (vector<[16]x[16]xi8>) -> ()
    return
  }
```

And `tileA` would still get spilled. Shouldn't spilling happen _after_ this "use"
```mlir
   "test.some_use"(%tileA) : (vector<[16]x[16]xi8>) -> ()
```
?

Reply:
Discussed offline.